keps/sig-network/0031-20181017-kube-proxy-services-optional.md

```diff
@@ -0,0 +1,135 @@
+---
+kep-number: 0
+title: Make kube-proxy service abstraction optional
+authors:
+  - "@bradhoekstra"
+owning-sig: sig-network
+participating-sigs:
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: "@bradhoekstra"
+creation-date: 2018-10-17
+last-updated: 2018-10-17
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Make kube-proxy service abstraction optional
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories](#user-stories)
+      * [Story 1](#story-1)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [Design](#design)
+      * [Considerations](#considerations)
+      * [Testing](#testing)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+
+## Summary
+
+In a cluster that has a service mesh a lot of the work being done by kube-proxy is redundant and wasted.
+Specifically, services that are only reached via other services in the mesh will never use the service abstraction implemented by kube-proxy in iptables (or ipvs).
+By informing the kube-proxy of this, we can lighten the work it is doing and the burden on its proxy backend.
+
+## Motivation
+
+The motivation for the enhancement is to allow higher scalability in large clusters with lots of services that are making use of a service mesh.
+
+### Goals
+
+The goal is to reduce the load on:
+* The apiserver sending all services and endpoints to all kube-proxy pods
+* The kube-proxy having to deserialize and process all services and endpoints
+* The backend system (e.g. iptables) for whichever proxy mode kube-proxy is using
+
+### Non-Goals
+
+* Making sure the service is still routable via the service mesh
+* Preserving any kube-proxy functionality for any intentionally disabled Service, including but not limited to: externalIPs, external LB routing, nodePorts, externalTrafficPolicy, healthCheckNodePort, UDP, SCTP
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster operator, operating a cluster using a service mesh I want to be able to disable the kube-proxy service implementation for services in that mesh to reduce overall load on the whole cluster
+
+### Implementation Details/Notes/Constraints
+
+#### Overview
+
+It is important for overall scalability that kube-proxy does not watch for service/endpoint changes that it is not going to affect. This can save a lot of load on the apiserver, networking, and kube-proxy itself by never requesting the updates in the first place. As such, annotating the services directly is considered insufficient as the kube-proxy would still have to watch for changed to the service.
```

Comment:
This won't save any load on API server, as far as I can see

Reply:
If all of the kube proxies aren't watching for Service and Endpoints objects for a namespace doesn't that lighten the load for the API server? Any updates that don't have to get sent to the kube-proxy should save some cycles for the API server.

Say there is a large service (>1000 pods) in a large cluster (>1000 nodes) that is constantly changing replica count because it is using HPA, that is a lot of large object updates that the API server doesn't have to send out to every node in the cluster.

Reply:
We don't watch individual namespaces, we watch all namespaces.  There's no server-side filtering that I know of for this.

@kubernetes/sig-api-machinery-feature-requests

Reply:
Watches can be cluster-wide or scoped to a single namespace (not a set of namespaces), and can be filtered by label selectors (or field selectors, depending on the object and fields in question)

Reply:
To be clear: the selector is on the leaf objects being watched, not on the namespaces being watched "into".

Reply:
If there is no server side filtering, it wont change any load at all 
@liggitt I believe everything you mentioned above is client side ?

Reply:
Filtering list and watch operations with label selectors is performed in the apiserver:
https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#list-and-watch-filtering